src/compiler/checker.ts

```diff
@@ -16976,20 +16978,20 @@ namespace ts {
                         if (maybeSuppress) {
                             overrideNextErrorInfo--;
                         }
-                        if (source.flags & TypeFlags.Object && target.flags & TypeFlags.Object) {
+                        if (sourceFlags & TypeFlags.Object && targetFlags & TypeFlags.Object) {

```

Comment:
This is one of the incorrectly captured flags since `target` might have changed from a union to an object:

```
                if (target.flags & TypeFlags.Union && source.flags & TypeFlags.Object &&
                    (target as UnionType).types.length <= 3 && maybeTypeOfKind(target, TypeFlags.Nullable)) {
                    const nullStrippedTarget = extractTypesOfKind(target, ~TypeFlags.Nullable);
                    if (!(nullStrippedTarget.flags & (TypeFlags.Union | TypeFlags.Never))) {
                        if (source === nullStrippedTarget) return Ternary.True;
                        target = nullStrippedTarget;
                    }
                }
```